### PR TITLE
Migrate read_file from pkg_resources to importlib.resources

### PR DESCRIPTION
**Pull request by eli-schwartz**
In python 3.9, we can get any resource from a package subdirectory using files(), but on older versions of python, even where importlib.resources exists, we need the backport.

Further progress on #4467

### CODE REVIEW

File: qutebrowser/utils/utils.py [hunk 1]
```diff
@@ -216,13 +220,12 @@ def read_file(filename: str, binary: bool = False) -> Any:
             with open(fn, 'r', encoding='utf-8') as f:
                 return f.read()
     else:
-        data = pkg_resources.resource_string(
-            qutebrowser.__name__, filename)
+        p = importlib_resources.files(qutebrowser) / filename
```
Comment (eli-schwartz) [inline qutebrowser/utils/utils.py:223]: Conveniently for line length, importlib_resources supports passing a module, not just a str.

File: qutebrowser/utils/utils.py [hunk 2]
```diff
@@ -42,6 +42,10 @@
 from PyQt5.QtCore import QUrl, QVersionNumber
 from PyQt5.QtGui import QClipboard, QDesktopServices
 from PyQt5.QtWidgets import QApplication
+if sys.version_info >= (3,9):
+    import importlib.resources as importlib_resources
+else:
+import importlib_resources
```
Comment (eli-schwartz) [inline qutebrowser/utils/utils.py:48]: What. How.

Comment (The-Compiler) [reply qutebrowser/utils/utils.py:48]: [Not your first PR](https://github.com/qutebrowser/qutebrowser/pull/5768/files#diff-a3f2f10f8adb634b3c7e2c6e18e468f4fea27185be7105f840d620783667a323R233) with a syntax error FWIW :wink:

File: qutebrowser/utils/utils.py [hunk 3]
```diff
@@ -42,6 +42,10 @@
 from PyQt5.QtCore import QUrl, QVersionNumber
 from PyQt5.QtGui import QClipboard, QDesktopServices
 from PyQt5.QtWidgets import QApplication
+if sys.version_info >= (3,9):
```
Comment (The-Compiler) [inline qutebrowser/utils/utils.py:45]:
```suggestion
if sys.version_info >= (3, 9):
```

Also, maybe add a comment mentioning that we need the `.files()` API which was added in 3.9 only?

File: setup.py [hunk 4]
```diff
@@ -71,7 +71,8 @@ def _get_constant(name):
         entry_points={'gui_scripts':
                       ['qutebrowser = qutebrowser.qutebrowser:main']},
         zip_safe=True,
-        install_requires=['pypeg2', 'jinja2', 'pygments', 'PyYAML', 'attrs'],
+        install_requires=['pypeg2', 'jinja2', 'pygments', 'PyYAML', 'attrs',
+                          'importlib_resources>=1.1.0; python_version < "3.9"'],
```
Comment (The-Compiler) [inline setup.py:75]: Please also add this to:

- The README
- `misc/requirements/requirements-qutebrowser.txt-raw`
- `_module_versions` in `qutebrowser/utils/version.py`
- and adjust `tests/unit/utils/test_version.py` accordingly (I can take care of that if you prefer)

**Review by The-Compiler — CHANGES_REQUESTED** [review]

File: setup.py (hunk 4 above)
Comment (eli-schwartz) [reply setup.py:75]: I'm guessing in requirements-qutebrowser.txt-raw it is unconditional?

File: qutebrowser/utils/utils.py [hunk 5]
```diff
@@ -42,6 +42,11 @@
 from PyQt5.QtCore import QUrl, QVersionNumber
 from PyQt5.QtGui import QClipboard, QDesktopServices
 from PyQt5.QtWidgets import QApplication
+# We cannot use the stdlib version on 3.7-3.8 because we need the files() API.
+if sys.version_info >= (3, 9):
+    import importlib.resources as importlib_resources
+else:
+    import importlib_resources
```
Comment (The-Compiler) [inline qutebrowser/utils/utils.py:49]: Feel free to ignore that one, I'll fix things up when regenerating the `requirements.txt`.

**Comment by The-Compiler** [comment]
Oh, wait - I thought I could take care of regenerating `requirements.txt`, but of course the tests won't run like this.

Here's the diff from running `python3 scripts/dev/recompile_requirements.py qutebrowser` and then removing the unrelated `Pygments` update:

```diff
diff --git i/requirements.txt w/requirements.txt
index fee8906ae..0e101a95d 100644
--- i/requirements.txt
+++ w/requirements.txt
@@ -2,6 +2,7 @@
 
 attrs==20.3.0
 colorama==0.4.4
+importlib-resources==3.3.0
 Jinja2==2.11.2
 MarkupSafe==1.1.1
 Pygments==2.7.2

```

could you please push that to your branch?

Also, note that I want to test if this works properly with PyInstaller on Windows/macOS, so it might take me another day or two until I get around to merging this.

**Comment by eli-schwartz** [comment]
> Missing changelog URLs in qutebrowser requirements: importlib-resources

**Comment by The-Compiler** [comment]
Ah, right, please also add a changelog URL to the giant dict of URLs in `scripts/dev/recompile_requirements.py`. Also note that `tests/unit/utils/test_version.py` seems to be failing.

**Comment by eli-schwartz** [comment]
> Also, note that I want to test if this works properly with PyInstaller on Windows/macOS, so it might take me another day or two until I get around to merging this.

OK -- the code anyways special cases pyinstaller / `if hasattr(sys, 'frozen'):` due to "PyInstaller doesn't support pkg_resources :(" so it would be interesting to see if this code fork could be removed (test via `if False:`) but otherwise should be a no-op.

**Comment by eli-schwartz** [comment]
Two failures:
- code not touched by this PR:
  > Warning: qutebrowser/utils/utils.py:77:4: C0412: Imports from package typing are not grouped (ungrouped-imports)
- code (probably?) touched by this PR:
  > Error: qutebrowser/utils/utils.py has 99.70% line and 99.19% branch coverage!

I've got no clue what codecov means by this and I don't believe in it anyway, so that can be someone else's problem.

Per https://github.com/qutebrowser/qutebrowser/issues/4467#issuecomment-741040644 the PyInstaller potential is a dead end. Anything else blocking this?

**Comment by The-Compiler** [comment]
Merged with a couple of small changes/fixes:

- 4a931f711 Update requirements.txt
- 57c514eb4 Fix pylint/coverage
- ed99cf123 Update docs

Thanks!

**Comment by The-Compiler** [comment]
(Also FWIW, that output was from `coverage.py`, not codecov.io if you mean that - it's also correct, on Python 3.9 where the coverage checking runs, the line importing the backport never runs)

**Comment by eli-schwartz** [comment]
Probably.

**Pull request merged**